### PR TITLE
Migrate comment fragments to Jetpack Compose

### LINKED ISSUE #10874: Introduce Jetpack Compose for NewPipe UI
### Checklist

- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have read the [FAQ](https://newpipe.net/FAQ/) and my problem isn't listed.
- [X] I'm aware that this is a request for NewPipe itself and that requests for adding a new service need to be made at [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor/issues).
- [X] I have taken the time to fill in all the required details. I understand that the feature request will be dismissed otherwise.
- [X] This issue contains only one feature request.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

#### Description:
- This ticket addresses the integration of Jetpack Compose as the UI framework for the NewPipe application.
- This will be base for the settings page redesign work being done in https://github.com/TeamNewPipe/NewPipe/issues/9587.

#### Tasks:
- Add necessary dependencies and setup for Jetpack Compose within the NewPipe project.
- Implement a starting composable component that adheres to NewPipe's design guidelines to demonstrate the integration.

#### Testing:
- Verify successful compilation and launch of the NewPipe application with Jetpack Compose integration.
- Confirm the demo composable renders with the appropriate design and theming.

#### Additional Notes:
- This ticket represents the initial phase of introducing Jetpack Compose.

### Why do you want this feature?

Given the planned NewPipe rewrite and requirement for redesign of settings page https://github.com/TeamNewPipe/NewPipe/issues/9587, this ticket will serve as base for the same.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Isira-Seneviratne**
#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

- Update the Compose dependencies to their latest versions.
- Add the [Compose library for Coil](https://coil-kt.github.io/coil/compose/) to handle image loading.
- Add the [LazyColumnScrollbar](https://github.com/nanihadesuka/LazyColumnScrollbar) library to display scrollbars in Compose.
- Migrate the comments to use Jetpack Compose and replace the comment replies fragment with a bottom sheet dialog.
- Handle loading the comments and comment replies using the [Paging](https://developer.android.com/jetpack/androidx/releases/paging) library and a [view model](https://developer.android.com/topic/libraries/architecture/viewmodel).

#### Before/After Screenshots/Screen Record

<table>
  <tr><th colspan=2>Before</th><tr>
  <tr>
    <th>Comments</th>
    <th>Comment replies</th>
  </tr>
  <tr>
    <td>Before</td>
    <td>Before</td>
  </tr>
</table>

<table>
  <tr><th colspan=3>After (initial design)</th><tr>
  <tr>
    <th>Comments</th>
    <th>Comment replies</th>
    <th>Expanded comment replies</th>
  </tr>
  <tr>
    <td>After</td>
    <td>After</td>
    <td>After</td>
  </tr>
</table>

<table>
  <tr><th colspan=3>After (improved design)</th><tr>
  <tr>
    <th>Comments</th>
    <th>Comment replies</th>
    <th>Expanded comment replies</th>
  </tr>
  <tr>
    <td>After</td>
    <td>After</td>
    <td>After</td>
  </tr>
</table>

https://github.com/TeamNewPipe/NewPipe/assets/31027858/72923272-bda6-4eb0-b1b4-63debae9d02d

#### Relies on the following changes

- #11238

#### APK testing

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Isira-Seneviratne** [comment]
> Thank you for the effort. Looks overall good to me. Very good UI and UX changes that make the app feel modern 👍
> 
> The light theme looks like this: 
> 
> Quick look, not a final review because I am currently lacking spare time

Fixed the issue, it was due to a recent change.

File: app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java [hunk 1]
```diff
@@ -1014,16 +1013,15 @@ public void scrollToTop() {
...
         // unexpand the app bar only if scrolling to the comment succeeded
-        if (((CommentsFragment) fragment).scrollToComment(comment)) {
-            binding.appBarLayout.setExpanded(false, false);
-            binding.viewPager.setCurrentItem(commentsTabPos, false);
-        }
+//        if (fragment instanceof CommentsFragment commentsFragment &&
+//                commentsFragment.scrollToComment(comment)) {
+//            binding.appBarLayout.setExpanded(false, false);
+//            binding.viewPager.setCurrentItem(commentsTabPos, false);
+//        }
```
[the remark this answers is not in the document]
Comment (Isira-Seneviratne) [reply app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:1024]: No, Compose seems to remember the current scrolled position when expanding/collapsing the player.

**Comment by tsiflimagas** [comment]
Hi, thank you so much for your work, it's terrific! Since your rewriting this part, do you think it would be a good chance to tackle #7191? Support in extractor has been added if I'm not wrong. Also, perhaps an m3 coloured scroll handle would blend in better. I would comment on a few more stuff, but they're things that were there before the migration (replies count on the bottom sheet header, copy on long press), so I guess you'll add them back eventually.

**Comment by Isira-Seneviratne** [comment]
> Hi, thank you so much for your work, it's terrific! Since your rewriting this part, do you think it would be a good chance to tackle #7191? Support in extractor has been added if I'm not wrong.

As far as I can tell, it hasn't been implemented yet, seeing as that issue is still open.

> Also, perhaps an m3 coloured scroll handle would blend in better.

I'm using [a third-party library](https://github.com/nanihadesuka/LazyColumnScrollbar) for the scrollbars, since first-party support [is yet to be implemented](https://developer.android.com/jetpack/androidx/compose-roadmap).

**Comment by tsiflimagas** [comment]
> we are happy about every reviewer, so if you have any comments on the PR leave them here😉

Thank you!

> As far as I can tell, it hasn't been implemented yet, seeing as that issue is still open.

The issue is open on the app side, since it hasn't been implemented in the app. If I'm not wrong [this PR](https://github.com/TeamNewPipe/NewPipeExtractor/pull/794) has added the necessary support in the extractor.

> I'm using a third-party library for the scrollbars, since first-party support is yet to be implemented.

By looking at the library's repo, it seemed to me that it offers the ability to edit scrollbars colour thus set it to an adaptive one, but maybe I got it wrong.

Also, it seems that none of the migrated fragments follows the black night theme. Perhaps it is planned to revamp app theming, or would it make sense to make these fragments follow app's theme as they it before?

**Comment by Isira-Seneviratne** [comment]
> > we are happy about every reviewer, so if you have any comments on the PR leave them here😉
> 
> Thank you!
> 
> > As far as I can tell, it hasn't been implemented yet, seeing as that issue is still open.
> 
> The issue is open on the app side, since it hasn't been implemented in the app. If I'm not wrong [this PR](https://github.com/TeamNewPipe/NewPipeExtractor/pull/794) has added the necessary support in the extractor.

Thanks, I missed that one.

> 
> > I'm using a third-party library for the scrollbars, since first-party support is yet to be implemented.
> 
> By looking at the library's repo, it seemed to me that it offers the ability to edit scrollbars colour thus set it to an adaptive one, but maybe I got it wrong.

Yeah, it's possible to change the scrollbar colour. How does red (the primary colour) sound?

> 
> Also, it seems that none of the migrated fragments follows the black night theme. Perhaps it is planned to revamp app theming, or would it make sense to make these fragments follow app's theme as they it before? 

Yeah, the black theme will need to be implemented.

**Comment by tsiflimagas** [comment]
> How does red (the primary colour) sound?

It's good I guess. The app hasn't been migrated to material You anyway, so it looks fitting. But you may choose what you find better, as the author of the PR.

**Comment by Stypox** [comment]
I pushed a few commits:
- redesigned the layouts, see comparison in PR body
- merge with `refactor` to include build fixes by litetex
- fix paging skipping first page of comments
- handle unknown values without showing -1 in various places
- you can test nested replies and unknown comment and like counts in this peertube video: https://share.tube/w/vxu4uTstUBAUromWwXGHrq

File: app/src/main/java/org/schabi/newpipe/ui/components/video/comment/CommentSection.kt [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                                        Comment(comment = comments[it]!!) {}
+                                    }
+                                }
+                            }
+                        }
+                    }
+
+                    is Resource.Error -> {
+                        item {
+                            NoItemsMessage(R.string.error_unable_to_load_comments)
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/ui/components/video/comment/CommentSection.kt:114]: An error panel should be created to handle all of these cases and allow retrying, reporting and opening in browser.

**Review by Stypox — APPROVED** [review]
Thank you! I implemented the TODOs in my previous comment.

The review comments I left here are things to do in future PRs.

**Review by AudricV — COMMENTED** [review]
Here is some general feedback to be added to my review comments:
- the difference between the two loading spinners is a bit disturbing, the new one for comments doesn't seem to match Material Design style;
- it would be great if the replies only prevent interacting with comments and not all content (i.e. comment drawers cannot be shown outside of the comments view).

**Comment by Profpatsch** [comment]
I created issues for the remaining post-review comments

**Comment by Stypox** [comment]
Now errors in the comments make the app crash completely:

## Exception
* __User Action:__ ui error
* __Request:__ ACRA report
* __Content Country:__ GB
* __Content Language:__ en-GB
* __App Language:__ en_GB
* __Service:__ none
* __Version:__ 0.27.2
* __OS:__ Linux Android 13 - 33
<details><summary><b>Crash log </b></summary><p>

```
java.lang.IllegalArgumentException: Page doesn't contain an URL
	at org.schabi.newpipe.extractor.services.soundcloud.extractors.SoundcloudCommentsExtractor.getPage(SoundcloudCommentsExtractor.java:44)
	at org.schabi.newpipe.extractor.comments.CommentsInfo.getMoreItems(CommentsInfo.java:79)
	at org.schabi.newpipe.paging.CommentsSource$load$info$1.invokeSuspend(CommentsSource.kt:23)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:101)
	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:113)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:89)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:589)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:823)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:720)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:707)
	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [LazyStandaloneCoroutine{Cancelling}@16d25a5, Dispatchers.Main.immediate]

```
</details>
<hr>

**Pull request merged**